sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/DownloadCBDE.targets

```diff
@@ -9,79 +9,93 @@
 
-    <PropertyGroup>
-      <PowerShellExe Condition="'$(PowerShellExe)'==''">%WINDIR%\System32\WindowsPowerShell\v1.0\powershell.exe</PowerShellExe>
-      <UnzipScript>$(MSBuildProjectDirectory)\..\..\..\Scripts\unzip.ps1</UnzipScript>
-      <DownloadScript>$(MSBuildProjectDirectory)\..\..\..\Scripts\download.ps1</DownloadScript>
-    </PropertyGroup>
-
-    <Message Importance="high" Text="CBDE: Begin downloading CBDE zip..." />
-    <Exec ConsoleToMsBuild="true"
-      Command="$(PowerShellExe) -NonInteractive -ExecutionPolicy Unrestricted -File $(DownloadScript) -url $(CBDEDownloadUrl) -outfolder $(CBDEFolder) -outfile $(CBDEZipFileName)">
-      <Output TaskParameter="ConsoleOutput" PropertyName="OutputOfExec" />
+    <Message Importance="high" Text="CBDE: Begin downloading CBDE zip..." Condition="$(ShouldDownloadCBDE)=='true'"/>
+    <Exec ConsoleToMsBuild="true" Condition="$(ShouldDownloadCBDE)=='true'"
+      Command="$(PowerShellExe) -NonInteractive -ExecutionPolicy Unrestricted -File $(DownloadScript) -url @(CBDEUrls) -outfolder $(CBDEFolder)%(CBDEUrls.SubFolder) -outfile %(CBDEUrls.ZipFileName)">
+      <Output TaskParameter="ConsoleOutput" PropertyName="OutputOfExec"/>
     </Exec>
-    <Message Importance="low" Text="Output of powershell download: $(OutputOfExec)" />
-    <Message Importance="high" Text="CBDE: Finished downloading CBDE binaries: $(CBDEZipFilePath)" />
+    <Message Importance="low" Text="Output of powershell download: $(OutputOfExec)"  Condition="$(ShouldDownloadCBDE)=='true'"/>
+    <Message Importance="high" Text="CBDE: Finished downloading CBDE binaries: $(CBDEZipFilePath)"  Condition="$(ShouldDownloadCBDE)=='true'"/>
 
-    <Message Importance="high" Text="CBDE: Unzipping the CBDE binaries..." />
-    <Exec ConsoleToMsBuild="true"
-      Command="$(PowerShellExe) -NonInteractive -ExecutionPolicy Unrestricted -File $(UnzipScript) -zipfile $(CBDEZipFilePath) -outpath $(CBDEFolder)">
+    <Message Importance="high" Text="CBDE: Unzipping the CBDE binaries..."  Condition="$(ShouldDownloadCBDE)=='true'"/>
+    <Exec ConsoleToMsBuild="true" Condition="$(ShouldDownloadCBDE)=='true'"
+      Command="$(PowerShellExe) -NonInteractive -ExecutionPolicy Unrestricted -File $(UnzipScript) -zipfile $(CBDEFolder)%(CBDEUrls.SubFolder)%(CBDEUrls.ZipFileName) -outpath $(CBDEFolder)%(CBDEUrls.SubFolder)">
       <Output TaskParameter="ConsoleOutput" PropertyName="OutputOfExec" />
     </Exec>
-    <Message Importance="low" Text="Output of powershell unzip: $(OutputOfExec)" />
+    <Message Importance="low" Text="Output of powershell unzip: $(OutputOfExec)" Condition="$(ShouldDownloadCBDE)=='true'"/>
 
-    <Message Importance="high" Text="CBDE: Finished unzipping CBDE binaries" />
+    <Message Importance="high" Text="CBDE: Finished unzipping CBDE binaries" Condition="$(ShouldDownloadCBDE)=='true'"/>
+
+    <!-- Specifies the CBDE files that should be included in the project and copied to the output directory -->
```

Comment:
This block was repeated in the prior version so it's not any worse than it was before, but it would be nice if it didn't have to be.
You could try a quick test of creating a new target containing just this `ItemGroup`, with the target specifying `AfterTargets="FetchCBDEBinaries"` and `BeforeTargets="BeforeBuild"`, but if you can't get it work easily I'd leave it as it is.

Reply:
It tried it, but if I understand you correctly, this target would have to be conditional, with the condition depending on the metadata of an item. And this is not allowed...

Reply:
It doesn't need to be conditional - the same lines are included unconditionally in 33-41.
The idea would be to have the ItemGroup in a target so that it is defined only once, and then make sure the target is run after the "download binaries" target.